File: DecibelMeter/Services/SoundLevelCalculator.swift
import Accelerate
import AVFAudio
import Foundation

enum SoundLevelCalculator {
    static let referenceOffset = 110.0
    static let minimumLevel = 0.0
    static let maximumLevel = 140.0

    static func decibels(rms: Double) -> Double {
        guard rms > 0 else { return -120 }
        return 20 * log10(rms)
    }

    static func estimatedLevel(dbFS: Double, calibrationOffset: Double) -> Double {
        min(
            max(dbFS + referenceOffset + calibrationOffset, minimumLevel),
            maximumLevel
        )
    }

    static func decibels(from buffer: AVAudioPCMBuffer) -> Double {
        guard
            let channelData = buffer.floatChannelData,
            buffer.frameLength > 0,
            buffer.format.channelCount > 0
        else {
            return -120
        }

        let frameCount = vDSP_Length(buffer.frameLength)
        var totalMeanSquare = Float.zero

        for channel in 0..<Int(buffer.format.channelCount) {
            var channelMeanSquare = Float.zero
            vDSP_measqv(channelData[channel], 1, &channelMeanSquare, frameCount)
            totalMeanSquare += channelMeanSquare
        }

        let meanSquare = totalMeanSquare / Float(buffer.format.channelCount)
        return decibels(rms: sqrt(Double(meanSquare)))
    }

    static func band(for level: Double) -> SoundLevelBand {
        switch level {
        case ..<55:
            return .low
        case ..<70:
            return .moderate
        case ..<85:
            return .high
        default:
            return .veryHigh
        }
    }
}
enum SoundLevelBand: String, CaseIterable {
    case low
    case moderate
    case high
    case veryHigh

    var title: String {
        switch self {
        case .low: "Low"
        case .moderate: "Moderate"
        case .high: "High"
        case .veryHigh: "Very high"
        }
    }

    var guidance: String {
        switch self {
        case .low:
            "A relatively quiet environment."
        case .moderate:
            "Noticeable sound; monitor comfort over time."
        case .high:
            "A loud environment; consider reducing exposure."
        case .veryHigh:
            "A very loud environment; move away or use appropriate protection."
        }
    }
}

File: DecibelMeter.Windows/Services/SoundLevelCalculator.cs
namespace DecibelMeter.Windows.Services
{
    /// <summary>
    /// Calculates sound levels in decibels and classifies into bands
    /// </summary>
    public static class SoundLevelCalculator
    {
        private const double ReferenceOffset = 110.0;
        public const double MinimumLevel = 0.0;
        public const double MaximumLevel = 140.0;

        /// <summary>
        /// Converts RMS value to decibels
        /// </summary>
        public static double Decibels(double rms)
        {
            if (rms <= 0)
                return -120;

            return 20 * Math.Log10(rms);
        }

        /// <summary>
        /// Converts raw dBFS to estimated sound pressure level
        /// </summary>
        public static double EstimatedLevel(double dbFS, double calibrationOffset)
        {
            var level = dbFS + ReferenceOffset + calibrationOffset;
            return Math.Clamp(level, MinimumLevel, MaximumLevel);
        }

        /// <summary>
        /// Calculates RMS from 16-bit audio samples
        /// </summary>
        public static double CalculateRmsFromInt16(short[] samples)
        {
            if (samples.Length == 0)
                return 0;

            double sum = 0;
            for (int i = 0; i < samples.Length; i++)
            {
                double normalized = samples[i] / 32768.0;
                sum += normalized * normalized;
            }

            double meanSquare = sum / samples.Length;
            return Math.Sqrt(meanSquare);
        }

        /// <summary>
        /// Calculates RMS from 32-bit float audio samples
        /// </summary>
        public static double CalculateRmsFromFloat(float[] samples)
        {
            if (samples.Length == 0)
                return 0;

            double sum = 0;
            for (int i = 0; i < samples.Length; i++)
            {
                sum += samples[i] * samples[i];
            }

            double meanSquare = sum / samples.Length;
            return Math.Sqrt(meanSquare);
        }

        /// <summary>
        /// Classifies a sound level into a band
        /// </summary>
        public static Models.SoundLevelBand GetBand(double level)
        {
            return level switch
            {
                < 55 => Models.SoundLevelBand.Low,
                < 70 => Models.SoundLevelBand.Moderate,
                < 85 => Models.SoundLevelBand.High,
                _ => Models.SoundLevelBand.VeryHigh
            };
        }
    }
}
